Submit/Evaluate txn result does not show error payloads

## Description

When submitting or evaluating a transaction in vscode, the returned value of the transaction is shown for success responses. For example:

```
[7/1/2019 4:16:57 PM] [INFO] evaluating transaction queryCar with args CAR001
[7/1/2019 4:16:57 PM] [SUCCESS] Returned value from queryCar: {"color":"red","make":"Ford","model":"Mustang","owner":"Brad"}
```

## Expected Behavior

I would also expect to see the payload for error responses.

## Actual Behavior

It looks like the message is shown for error responses, but not the payload. For example:

```
[7/1/2019 4:13:18 PM] [INFO] evaluating transaction queryCar with args CARX
[7/1/2019 4:13:18 PM] [ERROR] Error evaluating transaction: No peers available to query. Errors: [{"message":"transaction returned with failure: Car CARX does not exist","stack":"Error: transaction returned with failure: Car CARX does not exist\n\tat sendProposal.self._endorserClient.processProposal (/Users/jt/.vscode/extensions/ibmblockchain.ibm-blockchain-platform-1.0.3/node_modules/fabric-network/node_modules/fabric-client/lib/Peer.js:140:36)\n\tat Object.onReceiveStatus (/Users/jt/.vscode/extensions/ibmblockchain.ibm-blockchain-platform-1.0.3/node_modules/grpc/src/client_interceptors.js:1207:9)\n\tat InterceptingListener._callNext (/Users/jt/.vscode/extensions/ibmblockchain.ibm-blockchain-platform-1.0.3/node_modules/grpc/src/client_interceptors.js:568:42)\n\tat InterceptingListener.onReceiveStatus (/Users/jt/.vscode/extensions/ibmblockchain.ibm-blockchain-platform-1.0.3/node_modules/grpc/src/client_interceptors.js:618:8)\n\tat callback (/Users/jt/.vscode/extensions/ibmblockchain.ibm-blockchain-platform-1.0.3/node_modules/grpc/src/client_interceptors.js:845:24)","status":500,"payload":{"type":"Buffer","data":[]},"peer":{"url":"grpc://localhost:17051","name":"peer0.org1.example.com:17051","options":{"grpc.max_receive_message_length":-1,"grpc.max_send_message_length":-1,"grpc.keepalive_time_ms":120000,"grpc.http2.min_time_between_pings_ms":120000,"grpc.keepalive_timeout_ms":20000,"grpc.http2.max_pings_without_data":0,"grpc.keepalive_permit_without_calls":1,"name":"peer0.org1.example.com:17051","grpc.ssl_target_name_override":"peer0.org1.example.com","grpc.default_authority":"peer0.org1.example.com"}},"isProposalResponse":true,"vslsStack":[{},{},{},{},{}]}]
```

## Steps to Reproduce

1. Modify a generated java contract to return an error response with payload using [ResponseUtils.newErrorResponse(java.lang.String message, byte[] payload)](https://fabric-chaincode-java.github.io/org/hyperledger/fabric/shim/ResponseUtils.html#newErrorResponse-java.lang.String-byte:A-)
2. Install/instantiate the contract
3. Submit/evaluate the transaction

--- a/client/src/fabric/FabricConnection.ts
+++ b/client/src/fabric/FabricConnection.ts
@@ -31,6 +31,7 @@ export abstract class FabricConnection {
 
     private discoveryAsLocalhost: boolean;
     private discoveryEnabled: boolean;
+    private knownChannels: any;
 
     constructor(connectionProfilePath: string, outputAdapter?: OutputAdapter) {
         this.gateway = new Gateway();
@@ -59,14 +60,29 @@ export abstract class FabricConnection {
     public async getAllChannelsForPeer(peerName: string): Promise<Array<string>> {
         // TODO: update this when not just using admin
         const peer: Client.Peer = this.gateway.getClient().getPeer(peerName);
-        const channelResponse: Client.ChannelQueryResponse = await this.gateway.getClient().queryChannels(peer);
-
-        const channelNames: Array<string> = [];
-        channelResponse.channels.forEach((channel: Client.ChannelInfo) => {
-            channelNames.push(channel.channel_id);
-        });
-
-        return channelNames.sort();
+        try {
+            const channelResponse: Client.ChannelQueryResponse = await this.gateway.getClient().queryChannels(peer);
+            const channelNames: Array<string> = channelResponse.channels.map((channel: Client.ChannelInfo) => channel.channel_id);
+            return channelNames.sort();
+        } catch (error) {
+            if (error.message && error.message.match(/access denied/)) {
+                // Not allowed to do this as we're probably not an administrator.
+                const channelNames: Array<string> = [];
+                for (const channelName in this.knownChannels) {
+                    const channel: any = this.knownChannels[channelName];
+                    const peers: any = channel.peers || {};
+                    const peerNames: string[] = Object.keys(peers);
+                    if (peerNames.indexOf(peerName) > -1) {
+                        channelNames.push(channelName);
+                    }
+                }
+                // If we found any channels, great - if not, then we should rethrow the error.
+                if (channelNames.length > 0) {
+                    return channelNames;
+                }
+            }
+            throw error;
+        }
     }
 
     public async getInstantiatedChaincode(channelName: string): Promise<Array<{ name: string, version: string }>> {
@@ -119,7 +135,7 @@ export abstract class FabricConnection {
 
         this.discoveryAsLocalhost = this.hasLocalhostURLs(connectionProfile);
         this.discoveryEnabled = true;
-
+        this.knownChannels = connectionProfile['channels'] || {};
         const options: GatewayOptions = {
             wallet: wallet,
             identity: identityName,

--- a/client/test/fabric/FabricConnection.test.ts
+++ b/client/test/fabric/FabricConnection.test.ts
@@ -71,6 +71,18 @@ describe('FabricConnection', () => {
                     url: 'grpc://localhost:7050'
                 }
             },
+            channels: {
+                'channel-from-the-ccp-no-peers': {
+
+                },
+                'channel-from-the-ccp': {
+                    peers: {
+                        'peer0.org1.example.com': {
+
+                        }
+                    }
+                }
+            },
             certificateAuthorities: {
                 'ca.org1.example.com': {
                     url: 'https://localhost:7054'
@@ -296,7 +308,7 @@ describe('FabricConnection', () => {
 
     describe('getAllChannelsForPeer', () => {
         it('should get all the channels a peer has joined', async () => {
-            const peerOne: fabricClient.Peer = new fabricClient.Peer('grpc://localhost:1454', { name: 'peerOne' });
+            const peerOne: fabricClient.Peer = new fabricClient.Peer('grpc://localhost:1454', { name: 'peer0.org1.example.com' });
 
             fabricClientStub.getPeersForOrg.returns([peerOne]);
 
@@ -306,9 +318,49 @@ describe('FabricConnection', () => {
 
             await fabricConnection.connect(mockWallet, mockIdentityName);
 
-            const channelNames: Array<string> = await fabricConnection.getAllChannelsForPeer('peerTwo');
+            const channelNames: Array<string> = await fabricConnection.getAllChannelsForPeer('peer0.org1.example.com');
 
             channelNames.should.deep.equal(['channel-one', 'channel-two']);
+        });
+
+        it('should use the connection profile for the list of channels if the peer says access is denied', async () => {
+            const peerOne: fabricClient.Peer = new fabricClient.Peer('grpc://localhost:1454', { name: 'peer0.org1.example.com' });
+
+            fabricClientStub.getPeersForOrg.returns([peerOne]);
+
+            fabricClientStub.queryChannels.rejects(new Error('blah access denied blah'));
+
+            await fabricConnection.connect(mockWallet, mockIdentityName);
+
+            const channelNames: Array<string> = await fabricConnection.getAllChannelsForPeer('peer0.org1.example.com');
+
+            channelNames.should.deep.equal(['channel-from-the-ccp']);
+        });
+
+        it('should rethrow the error if the peer says access is denied and there are no matching channels in the connection profile', async () => {
+            const peerOne: fabricClient.Peer = new fabricClient.Peer('grpc://localhost:1454', { name: 'peer0.org2.example.com' });
+
+            fabricClientStub.getPeersForOrg.returns([peerOne]);
+
+            fabricClientStub.queryChannels.rejects(new Error('blah access denied blah'));
+
+            await fabricConnection.connect(mockWallet, mockIdentityName);
+
+            await fabricConnection.getAllChannelsForPeer('peer0.org2.example.com')
+                .should.be.rejectedWith(/blah access denied blah/);
+        });
+
+        it('should rethrow any other error', async () => {
+            const peerOne: fabricClient.Peer = new fabricClient.Peer('grpc://localhost:1454', { name: 'peer0.org1.example.com' });
+
+            fabricClientStub.getPeersForOrg.returns([peerOne]);
+
+            fabricClientStub.queryChannels.rejects(new Error('such error'));
+
+            await fabricConnection.connect(mockWallet, mockIdentityName);
+
+            await fabricConnection.getAllChannelsForPeer('peer0.org1.example.com')
+                .should.be.rejectedWith(/such error/);
         });
     });
 

